CorsixTH/Lua/utility.lua

```diff
@@ -311,3 +311,37 @@ function rangeMapLookup(number, buckets)
   end
   assert(false) -- Should never get here.
 end
+
+if _G._VERSION == "Lua 5.3" then
+  -- use builtins 5.3
+  function bitOr(value1, value2)
+    return value1 | value2
+  end
+
+  function hasBit(value, bit)
+    return (value & 2^bit) ~= 0
+  end
+elseif _G._VERSION == "Lua 5.2" then
+  -- use builtins 5.2
+  function bitOr(value1, value2)
+    return bit32.bor(value1, value2)
+  end
+
+  function hasBit(value, bit)
+    return bit32.btest(value, 2^bit)
+  end
+else
+  -- this is a pseudo bitwise OR operation
+  -- assumes value2 is always a power of 2 (limits carry errors in the addition)
+  -- mimics the logic of hasBit with the addition if bit not set
+  function bitOr(value1, value2)
+    return value1 % (value2 + value2) >= value2 and value1 or value1 + value2
+  end
+  --! Check bit is set
+  --! param value - value to check set bit of
```

Comment:
Not sure if it's the cause of the error, but there should be no space between the `!` and `param` here. See https://github.com/CorsixTH/CorsixTH/wiki/Coding-Conventions#ldocgen-syntax

Reply:
I have 5.3 supported syntax for the bitwise operators in utility.lua - specifically bitOr and the value1 | value2
lua_tokenise needs a slight alteration to include them 
operator = Ct(C(P".." + P"<=" + P">=" + P"==" + P"~=" + P">>" + P"<<" + S"+-*/^%<>#&|~") * Cc"operator"),

Travis still fails as it doesn't use lua 5.3 (yet) - luacheck on utility.lua fails for the same reason, thats not too much more work to build lua in the travis build.

Reply:
Thanks for looking into it! Those changes seem like a prerequisite for this PR, as long as you want to continue using lua 5.3 bitwise operators.

It's unfortunate that we can't continue to automatically build and test lua 5.2, as I'm sure compatibility will degrade if it's not used. That said the situations where the latest CorsixTH was available, but only lua 5.2 are likely very close to 0 these days.

Reply:
Ok, I see the issue now. Just my misunderstanding.

I could do something like the following for those functions and as long as 5.3 was compiled with compatability on it would use the bit32 stuff, else fall back to the old stuff or just use the old stuff (which is probably the way to go). Just that bitOr routine isn't a real Or - it Ors a single bit only effectively

bitOr = bit32 and function(value1, value2) return bit32.bor(value1, value2) end or function(value1, value2)
  return value1 % (value2 + value2) >= value2 and value1 or value1 + value2
end

hasBit = bit32 and function(value, bit) return bit32.btest(value, 2^bit) end or function (value, bit)
  local p = 2 ^ bit
  return value % (p + p) >= p
end